include/sparrow/allocator.hpp

```diff
@@ -0,0 +1,247 @@
+        
+        any_allocator select_on_container_copy_construction() const;
+
+        bool equal(const any_allocator& rhs) const; 
+
+    private:
+
+        struct interface
+        {
+            [[nodiscard]] virtual T* allocate(std::size_t) = 0;
+            virtual void deallocate(T*, std::size_t) = 0;
+            virtual std::unique_ptr<interface> clone() const = 0;
+            virtual bool equal(const interface&) const = 0;
+            virtual ~interface() = default;
+        };
+
+        template <allocator A>
+        struct impl : interface
+        {
+            A m_alloc;
+
+            explicit impl(A alloc) : m_alloc(std::move(alloc)) {}
+
+            [[nodiscard]] T* allocate(std::size_t n) override
+            {
+                return m_alloc.allocate(n);
+            }
+
+            void deallocate(T* p, std::size_t n) override
+            {
+                m_alloc.deallocate(p, n);
+            }
+
+            std::unique_ptr<interface> clone() const override
+            {
+                return std::make_unique<impl<A>>(m_alloc);
+            }
+
+            bool equal(const interface& rhs) const override
+            {
+                if (auto* p = dynamic_cast<const impl<A>*>(&rhs))
+                {
+                    return p->m_alloc == m_alloc;
+                }
+                return false;
+            }
+        };
+
+        using storage_type = std::variant
+        <
+            std::allocator<T>,
+            std::pmr::polymorphic_allocator<T>,
+            std::unique_ptr<interface>
+        >;
+
+        template <class A>
+        std::unique_ptr<interface> make_storage(A&& alloc) const
+        {
+            return std::make_unique<impl<std::decay_t<A>>>(std::forward<A>(alloc));
+        }
+
+        template <class A>
+        requires is_value_stored<A>
+        A&& make_storage(A&& alloc) const
+        {
+            return std::forward<A>(alloc);
+        }
+
+        storage_type copy_storage(const storage_type& rhs) const
+        {
+            return std::visit([](auto&& arg) -> storage_type
+            {
+                using A = std::decay_t<decltype(arg)>;
+                if constexpr (is_value_stored<A>)
+                    return { A(arg) };
+                else
+                    return { arg->clone() };
+            }, rhs);
+        }
+
+        storage_type m_storage;
+    };
+
+    /********************************
+     * any_allocator implementation *
+     ********************************/
+
+    template <class T, class DA>
+    any_allocator<T, DA>::any_allocator()
+        : m_storage(make_storage(DA()))
+    {
+    }
+
+    template <class T, class DA>
+    any_allocator<T, DA>::any_allocator(const any_allocator& rhs)
+        : m_storage(copy_storage(rhs.m_storage))
+    {
+    }
+
+    template <class T, class DA>
+    [[nodiscard]] T* any_allocator<T, DA>::allocate(std::size_t n)
+    {
+        return std::visit([n](auto&& arg)
+        {
```

Comment:
looks like thius lambda is repeated partially, maybe ther eis a way to factorize this?

Reply:
In that case I don't see any mean to factorize the code without making it overcomplicated: you would need to pass two additional lambdas to a common function. I find it a bit of overkill to factorize out a simple if.

Reply:
The only thing that change in these function is the syntax to access the stored object. So a general accessor which provide a common interface whatever the type would solve the issue. Something to be able to change this code to
```c++
return some_nice_name([n](auto& allocator){ return allocator.allocate(n); };
```

I disagree that it's overcomplicated (to have that accessor/mutator/visitor function) because I find the current access code repeated obfuscating the important part of these functions: we just want to call allocate and deallocate here, whatever the type.

Reply:
> I disagree that it's overcomplicated

Indeed, what I had in mind was factorizing the control flow in a third function that would be called by these two methods. Providing accessors sohuld actually be quite trivial

Reply:
> I disagree that it's overcomplicated

Indeed, it's not the "design" I had in mind when I commented. Factorieing the code with a general accessor should be easier.